File: apps/web/src/shared/api/problem.ts
// RFC 9457 Problem Details parsing. docs/api/openapi.yaml `Problem`.
import type { components } from '@nullnull/api-client/types';

export type Problem = components['schemas']['Problem'];
export type ProblemCode = Problem['code'];
export type FieldError = components['schemas']['FieldError'];

/** Every code in the contract enum, in the order docs/api/README.md lists them. */
export const PROBLEM_CODES = [
  'INVALID_REQUEST',
  'UNAUTHORIZED',
  'FORBIDDEN',
  'NOT_FOUND',
  'VALIDATION_FAILED',
  'CSRF_INVALID',
  'CURSOR_INVALID',
  'CURSOR_EXPIRED',
  'TRIP_CHANGED',
  'DATA_CHANGED',
  'LOCK_CONFLICT',
  'ROUTE_UNAVAILABLE',
  'NO_IMPROVEMENT',
  'APPLY_FAILED',
  'IDEMPOTENCY_KEY_REUSED',
  'IMPORT_DRAFT_EXPIRED',
  'IMPORT_DRAFT_CHANGED',
  'PREVIEW_EXPIRED',
  'REVERT_WINDOW_EXPIRED',
  'DELETION_STATUS_EXPIRED',
  'SOURCE_UNAVAILABLE',
  'RATE_LIMITED',
  'INTERNAL_ERROR',
] as const satisfies readonly ProblemCode[];

const KNOWN_CODES = new Set<string>(PROBLEM_CODES);

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}

/**
 * True only for a body that carries every field the contract requires.
 *
 * A body with an unrecognised `code` is rejected. The server may add codes
 * ahead of the client, but surfacing one the policy table has never seen would
 * mean guessing its retry and recovery behaviour — the caller handles that as
 * an unrecognised failure instead.
 */
export function isProblem(value: unknown): value is Problem {
  if (!isRecord(value)) return false;
  return (
    typeof value.type === 'string' &&
    typeof value.title === 'string' &&
    typeof value.status === 'number' &&
    typeof value.code === 'string' &&
    KNOWN_CODES.has(value.code) &&
    typeof value.detail === 'string' &&
    typeof value.instance === 'string' &&
    typeof value.requestId === 'string' &&
    typeof value.retryable === 'boolean'
  );
}

/**
 * Narrow a failed request to a Problem, or null when it is not one.
 *
 * Returns null for network failures, HTML error pages from a proxy, and any
 * body that is not shaped like a Problem. Those are real failures but they are
 * a different kind: inventing a code for them would put a value on screen that
 * the contract never defined (docs/api/README.md forbids exposing anything as a
 * new code). Callers must handle null as "something failed, cause unknown".
 *
 * `response` is optional so this also works on an openapi-fetch `error` value
 * alone; when given, a non-problem+json content type is rejected outright.
 */
export function toProblem(error: unknown, response?: Response): Problem | null {
  if (response) {
    const contentType = response.headers.get('content-type') ?? '';
    if (!contentType.includes('application/problem+json')) return null;
  }
  return isProblem(error) ? error : null;
}

File: apps/api/src/main/java/io/nullnull/shared/problem/Problem.java
package io.nullnull.shared.problem;

import com.fasterxml.jackson.annotation.JsonInclude;
import java.util.List;
import java.util.Objects;

/**
 * RFC 9457 Problem Details with the stable Nullnull {@code code}. Field set and nullability follow
 * {@code Problem} in docs/api/openapi.yaml: the first eight fields are required, the optional
 * fields are omitted from JSON when null so that {@code additionalProperties: false} and the
 * non-nullable {@code fieldErrors} array type both hold.
 */
public record Problem(
        String type,
        String title,
        int status,
        ProblemCode code,
        String detail,
        String instance,
        String requestId,
        boolean retryable,
        @JsonInclude(JsonInclude.Include.NON_NULL) List<FieldError> fieldErrors,
        @JsonInclude(JsonInclude.Include.NON_NULL) Long currentTripVersion,
        @JsonInclude(JsonInclude.Include.NON_NULL) String recomputeUrl) {

    public static final String MEDIA_TYPE = "application/problem+json";

    public Problem {
        Objects.requireNonNull(type, "type");
        Objects.requireNonNull(title, "title");
        Objects.requireNonNull(code, "code");
        Objects.requireNonNull(detail, "detail");
        Objects.requireNonNull(instance, "instance");
        Objects.requireNonNull(requestId, "requestId");
        if (status < 400 || status > 599) {
            throw new IllegalArgumentException("problem status must be 4xx or 5xx: " + status);
        }
        fieldErrors = fieldErrors == null ? null : List.copyOf(fieldErrors);
    }

    public static Problem of(ProblemCode code, int status, String detail, String instance,
            String requestId, boolean retryable) {
        return new Problem(code.typeReference(), code.title(), status, code, detail, instance,
                requestId, retryable, null, null, null);
    }

    public Problem withFieldErrors(List<FieldError> errors) {
        return new Problem(type, title, status, code, detail, instance, requestId, retryable,
                errors, currentTripVersion, recomputeUrl);
    }
}
